### DISCUSSION TITLE
How to place footer on last page without leaving gap on rest of the pages

### DISCUSSION THREAD

**Discussion by matewins** [eclipse-birt/birt#959] [Q&A]
Hello,

I've been tasked with converting some reports over to Birt but with the stipulation that a fixed size footer only appears on the last page. Following suggestions online, that seemed relatively straightforward by adding a conditional to the contents of the footer at the _onRender_ event to check the _pageNumber_ variable and show or hide the contents accordingly. However this approach still leaves a gap on all the previous pages which I've been asked to look into removing.

Is there any way to remove these gaps and just have a footer at the very bottom of the last page?

I would imagine Birt needs to have some idea of the size of the page when it's constructing the document, but I was wondering if I could move the contents of our footer (just a grid with some summary information) out of the Master Page and into the main layout and anchor that to the bottom of the last page somehow. It doesn't look like this is possible via the standard styling options, but I wandered if I could dynamically alter the top margin of the contents of the "footer" such that it would resize itself to fill in the gap based on where the preceding element was rendered on the page.

I've tried to access the methods _getVerticalPosition_ and _getHorizontalPosition_ via the _onRender_ event but these all seem to just return null for every element. Is it possible to get any information about the layout of the document as it is being constructed and access that via the scripting tools? Or have a misunderstood exactly when and what is invoking the _onRender_ functions? Is there some other means to achieve a footer on the last page without leaving a gap at the bottom of the other page?

Thanks!
Mat

**Comment by hvbtup** [comment]
> Is there any way to remove these gaps and just have a footer at the very bottom of the last page?

No.

But the branch tup_main of my fork hvbtup/birt on GH contains a feature which can help to achieve this effect with the PDF emitter.

So if you feel brave and your goal is to create a PDF-only report, you can test if this works for you and we could create a PR to add this feature to the master branch of eclipse/birt.

You can imagine this like a vertical tab stop.

The commit was https://github.com/hvbtup/birt/commit/cc8302352ca3e6d8ac09ee87866d58b3f5692fac

With this code, you can add a UserProperty called "VerticalTab" with a string value like "180mm" to a cell of a grid/table,
or a UserProperty called "FixYPosition" to a dynamic text item or data item.

This works more or less like you describe.

I have to admit that I developed this because I was told a client needed this. Later on it turned out that this was not a strict requirement, so we didn't use this in production reports. I nevertheless kept the feature because I thought it could still be useful.

I know that I created a simple demo report for this feature, but it is not tested in real-world reports.
Unfortunately, I recently switched hardware and it seems that I didn't keep the demo report, so I can't attach it here.

Note: Of course you would'nt use this UserProperty in a master page footer.
Instead, leave your master page footer empty (or in such a way that it's the same for all pages) and use the property for a layout item inside the body.

See https://github.com/hvbtup/birt/wiki/NotableDifferences for more info about the fork (might be outdated regarding the bugs).

**Reply by matewins** [reply]
Yeah that looks exactly like what we're trying to achieve, "vertical tab stop" is an excellent way of putting it.

I was hoping the ability to alter the layout in this fashion might be accessible via the 'script' functions - only because the Birt runtime that we use to generate these reports is part of some third party commercial software so I need to check how we would go about deploying any changes to production at our end (at least for the period until these features were incorporated into master and vendor gets around to updating the runtime at their end) - but there's some very useful suggestions here to have a play around with.

Thanks very much.

**Reply by hvbtup** [reply]
It would definitely be much easier if one could access the "current y position" from a script. My first attempts where more or less the same what you did. But just like you, I didn't find a way to get this information from a script.

Then I thought that maybe there is a good reason for this:
If one could access the current position while the rendering task is already running, this would influence the rendering task, causing something like a feedback loop which in turn could result in hard-to-find bugs.

Personally, I don't really understand how the whole rendering process is working internally. I tried to understand this by looking at the code, but it is far too deeply nested for my brain.

Nevertheless I tried to find a way to reach my goal at least for the PDF emitter, so I debugged a simple report to see which packages are used, then looked for promising variable names in their code, added some debug statements here and there etc. which finally resulted in that commit.

If you are using BIRT as part of a commercial software, then maybe the vendor can create a PR for BIRT based on my commit.
